hub/core/chunk_engine/write.py

```diff
@@ -0,0 +1,167 @@
+import numpy as np
+import pickle
+from uuid import uuid1
+
+from hub.core.chunk_engine import generate_chunks
+
+from hub.core.typing import Provider
+from typing import Any, Callable, List, Tuple
+
+from .util import (
+    array_to_bytes,
+    normalize_and_batchify_shape,
+    get_meta_key,
+    get_index_map_key,
+    get_chunk_key,
+)
+
+
+def write_array(
+    array: np.ndarray,
+    key: str,
+    chunk_size: int,
+    storage: Provider,
+    batched: bool = False,
+    tobytes: Callable[[np.ndarray], bytes] = array_to_bytes,
+):
+    """Chunk & write an array to storage.
+
+    For more on chunking, see the `generate_chunks` method.
+
+    Args:
+        array (np.ndarray): Array to be chunked/written. Batch axis (`array.shape[0]`) is optional, if `array` does have a
+            batch dimension, you should pass the argument `batched=True`.
+        key (str): Key for where the chunks, index_map, & meta will be located in `storage` relative to it's root.
+        chunk_size (int): Desired length of each chunk.
+        storage (Provider): Provider for storing the chunks, index_map, & meta.
+        batched (bool): If True, the provied `array`'s first axis (`shape[0]`) will be considered it's batch axis.
+            If False, a new axis will be created with a size of 1 (`array.shape[0] == 1`). default=False
+        tobytes (Callable): Must accept an `np.ndarray` as it's argument & return `bytes`.
+
+    Raises:
+        NotImplementedError: Do not use this function for writing to a key that already exists.
+    """
+
+    array = normalize_and_batchify_shape(array, batched=batched)
+
+    meta_key = get_meta_key(key)
+    index_map_key = get_index_map_key(key)
+    if meta_key in storage or index_map_key in storage:
+        raise NotImplementedError("Appending is not supported yet.")
+
+    index_map: List[dict] = []
+    meta = {
+        "chunk_size": chunk_size,
+        "dtype": array.dtype.name,
+        "length": array.shape[0],
+        "min_shape": array.shape[1:],
+        "max_shape": array.shape[1:],
+    }
+
+    for i in range(array.shape[0]):
+        sample = array[i]
+        b = tobytes(sample)
+
+        index_map_entry = write_bytes(b, key, chunk_size, storage, index_map)
+
+        # shape per sample for dynamic tensors (TODO: if strictly fixed-size, store this in meta)
+        index_map_entry["shape"] = sample.shape
+        index_map.append(index_map_entry)
+
+    # TODO: don't use pickle
```

Comment:
Why not just... not use it? What's the secret here?

Reply:
because there's no point in storing it more efficiently if this is only a temporary index_map. meta/index_map should be in a separate PR.